File: android/src/main/java/com/easemob/im_flutter_sdk/ConversationWrapper.java
package com.easemob.im_flutter_sdk;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;

import com.hyphenate.chat.EMClient;
import com.hyphenate.chat.EMConversation;
import com.hyphenate.chat.EMMessage;

import org.json.JSONException;
import org.json.JSONObject;


public class ConversationWrapper extends Wrapper implements MethodCallHandler{

    ConversationWrapper(FlutterPlugin.FlutterPluginBinding flutterPluginBinding, String channelName) {
        super(flutterPluginBinding, channelName);
    }

    @Override
    public void onMethodCall(MethodCall call, Result result) {

        JSONObject param = (JSONObject)call.arguments;

        try { 
            if (MethodKey.getLatestMessage.equals(call.method)) {
                getLatestMessage(param, call.method, result);
            }
            else if (MethodKey.getLatestMessageFromOthers.equals(call.method)) {
                getLatestMessageFromOthers(param, call.method, result);
            }

            else if (MethodKey.conversationDeleteServerMessageWithTime.equals(call.method)) {
                deleteLocalAndServerMessagesByTime(param, call.method, result);
            }
            else
            {
                super.onMethodCall(call, result);
            }
        } catch (JSONException e) {
            e.printStackTrace();
        }
    }

    private void getLatestMessage(JSONObject params, String channelName, Result result) throws JSONException {
        EMConversation conversation = conversationWithParam(params);

        asyncRunnable(()->{
            EMMessage msg = conversation.getLastMessage();
            onSuccess(result, channelName, msg != null ? MessageHelper.toJson(msg) : null);
        });
    }

    private void getLatestMessageFromOthers(JSONObject params, String channelName, Result result) throws JSONException {
        EMConversation conversation = conversationWithParam(params);

        asyncRunnable(()->{
            EMMessage msg = conversation.getLatestMessageFromOthers();
            onSuccess(result, channelName, msg != null ? MessageHelper.toJson(msg) : null);
        });
    }

    private EMConversation conversationWithParam(JSONObject params ) throws JSONException {
        String convId = params.getString("convId");
        EMConversation.EMConversationType type = EnumTools.conversationTypeFromInt(params.getInt("type"));
        return EMClient.getInstance().chatManager().getConversation(convId, type, true);
    }

    private void deleteLocalAndServerMessagesByTime(JSONObject params, String channelName, Result result) throws JSONException {
        EMConversation conversation = conversationWithParam(params);
        long beforeMs = params.optLong("beforeMs");
        conversation.removeMessagesFromServer(beforeMs, new EMWrapperCallBack(result, channelName,null));
    }

}

File: ios/Classes/ConversationWrapper.m
//
//  EMConversationWrapper.m
//  
//
//  Created by 杜洁鹏 on 2019/10/8.
//

#import "ConversationWrapper.h"
#import "MethodKeys.h"

#import "MessageHelper.h"
#import "ConversationHelper.h"
#import "EnumTools.h"
#import "Helper.h"

@interface ConversationWrapper ()

@end

@implementation ConversationWrapper
- (instancetype)initWithChannelName:(NSString *)aChannelName
                          registrar:(NSObject<FlutterPluginRegistrar>*)registrar {
    if(self = [super initWithChannelName:aChannelName
                               registrar:registrar]) {
    }
    return self;
}

#pragma mark - FlutterPlugin

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
     if ([ChatGetLatestMsg isEqualToString:call.method]) {
        [self getLatestMessage:call.arguments
                   channelName:call.method
                        result:result];
    } else if ([ChatGetLatestMsgFromOthers isEqualToString:call.method]) {
        [self getLatestMessageFromOthers:call.arguments
                             channelName:call.method
                                  result:result];
    } else if ([ChatConversationDeleteServerMessageWithTime isEqualToString:call.method]) {
        [self deleteServerMessagesByTime:call.arguments
                         channelName:call.method
                              result:result];
    }
    
    
    
    else {
        [super handleMethodCall:call result:result];
    }
}


#pragma mark - Private
- (void)getConversationWithParam:(NSDictionary *)param
                      completion:(void(^)(EMConversation *conversation))aCompletion
{
    __weak NSString *conversationId = param[@"convId"];
    EMConversationType type = [EnumTools conversationTypeFromInt:[param[@"type"] intValue]];
    EMConversation *conversation = [EMClient.sharedClient.chatManager getConversation:conversationId
                                                                                 type:type
                                                                     createIfNotExist:YES];
    if (aCompletion) {
        aCompletion(conversation);
    }
}

#pragma mark - Actions

- (void)getLatestMessage:(NSDictionary *)param
             channelName:(NSString *)aChannelName
                  result:(FlutterResult)result
{
    __weak typeof(self) weakSelf = self;
    [self getConversationWithParam:param
                        completion:^(EMConversation *conversation)
     {
        EMChatMessage *msg = conversation.latestMessage;
        [weakSelf wrapperCallBack:result
                      channelName:aChannelName
                            error:nil
                           object:[msg toJson]];
    }];
}

- (void)getLatestMessageFromOthers:(NSDictionary *)param
                       channelName:(NSString *)aChannelName
                            result:(FlutterResult)result
{
    __weak typeof(self) weakSelf = self;
    [self getConversationWithParam:param
                        completion:^(EMConversation *conversation)
     {
        EMChatMessage *msg = conversation.lastReceivedMessage;
        [weakSelf wrapperCallBack:result
                      channelName:aChannelName
                            error:nil
                           object:[msg toJson]];
    }];
}


- (void)deleteServerMessagesByTime:(NSDictionary *)param
                   channelName:(NSString *)aChannelName
                      result:(FlutterResult)result{
    long long ts = [param[@"beforeMs"] longLongValue];
    __weak typeof(self) weakSelf = self;
    [self getConversationWithParam:param
                        completion:^(EMConversation *conversation)
     {
        [conversation removeMessagesFromServerWithTimeStamp:ts completion:^(EMError * _Nullable aError) {
            [weakSelf wrapperCallBack:result
                          channelName:aChannelName
                                error:aError
                               object:nil];
        }];
    }];
}


@end
